app/controllers/results_controller.rb

```diff
@@ -166,11 +160,22 @@ def show
                                  peer_visible: is_review)
           criteria_info = criteria.pluck_to_hash(*fields)
           marks_info = criteria.joins(:marks)
-                               .where('marks.result_id': result.id)
-                               .pluck_to_hash(*fields, 'marks.mark')
-                               .group_by { |h| h[:id] }
-          criteria_info.map do |h|
-            info = marks_info[h[:id]]&.first || h.merge('marks.mark': nil)
+                              .where('marks.result_id': result.id)
+                              .pluck_to_hash(*fields, 'marks.mark')
+                              .group_by { |h| h[:id] }
+          # adds an extra levels field to the marks info hash with the same rubric criterion id
+          levels_info = []
+          if klass == RubricCriterion
+            criteria_info.map do |cr|
+              levels_info = Level.where(rubric_criterion_id: cr[:id])
+                                  .pluck_to_hash(:name, :description, :mark)
+              rubric_marks = marks_info[cr[:id]]&.first
+              rubric_marks.merge!(levels: levels_info)
```

Comment:
Note that `&.` (on the previous line) can return `nil`. So you can just do this call in one line, e.g. `...&.first&.merge!(levels: ...)` to avoid having a temp variable `rubric_marks`.